### ISSUE TITLE
RFC: `unless` doesn't support hypothesis like e.g. `if h: ...`

### ISSUE THREAD

**Issue by TomasPuverle** [leanprover/lean4#5598]
### Proposal

A lot of (most?) control statements support a hypothesis on the condition they evaluate, however `unless` does not.  It would be useful to add this, esp. in light of my other proposal/PR about extending the hypothesis scope: #5585.

Testing a condition and taking an early return is a very common operation and Lean should support writing the code as "naturally" as possible.

  - **User Experience**
(Especially in combination) The above two features would result in being able to write code like this:

```lean
def getFirst(arr : Array Nat) : Option Nat :=
   unless h: arr1.size > 0 return none
   return arr[0]   -- Note no []! index needed
```

  - **Beneficiaries**: Which Lean users and projects benefit most from this feature/change?

This seems to be a fairly common pattern in the Lean source code e.g. in `BrecOn.lean` in `replaceRecApps`

```lean
                unless xs.size = numParams do
                  throwError "unexpected matcher application alternative{indentExpr alt}\nat application{indentExpr e}"
                let belowForAlt := xs[numParams - 1]!       -- Note the []! here
```

or `InlineCandidate.lean` in `inlineCandidate?`:

```lean
      unless paramIdx < args.size do return none
      let arg := args[paramIdx]!
```

  - **Maintainability**: Will this change streamline code maintenance or simplify its structure?

As demonstrated by the examples directly from the Lean's source.

### Community Feedback

I have brought this up on Zulip previously and it has resurfaced [here](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/RFC.3A.20If.20hypothesis.20scope/near/474346059)

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others benefit from the changes in this proposal being added, please ask them to add :+1: to it.

**Comment by Kha** [comment]
RFC accepted; the design of the new syntax is fine but note that regarding an implementation, the `do` elaborator code is currently frozen

**Comment by digama0** [comment]
What does that mean, concretely? Assuming that it means that FRO people will likely not work on it and external PRs will not be considered, I'm not sure how this differs from not being accepted at all.

**Comment by nomeata** [comment]
It means that the feature is desirable in principle, and when and if the do notation modules is being rewritten we hope we can close this rfc as implemented (or then accept a PR once the code is in a shape where that makes sense)

**Comment by eric-wieser** [comment]
Here's a hacky implementation:
```lean
import Lean
open Lean Parser Term

/-- `unless h : P do body` is like `unless P do body`, but makes `h` available afterwards. -/
@[doElem_parser] def doUnlessHyp := leading_parser
  "unless " >> Term.ident >> ":" >> withForbidden "do" termParser >> " do " >> doSeq

macro_rules
| `(doElem| unless $h : $P do $body) =>
  `(doElem| let ⟨$h⟩ <- if h : $P then pure <| PLift.up h else $body)


def getFirst(arr : Array Nat) : Option Nat := do
   unless h : arr.size > 0 do failure
   return arr[0]
```

**Comment by l1mey112** [comment]
Bump

**Comment by sgraf812** [comment]
Consider

```lean
/-- info: 42 -/
#guard_msgs in
#eval do
  let i := 42
  unless i < 42 do
    pure ()
  return i
```

What should happen if the user changes this to

```lean
#eval do
  let i := 42
  unless h : i < 42 do
    pure ()
  return i
```

? This is to say: We are talking about an entirely different semantics for `unless` depending on whether or not `h : ` is present. That seems really confusing.

Maybe you are saying that `unless h : i < 42 do c` should only be allowed if `c` is not falling through (e.g. always `return`s). That might be sensible, as in that case one could expand `do unless h : p do c; rest` to `do if h : p then rest else c; unreachable!`. 

But then we should maybe also allow arbitrary `doIfCond`s such as `unless let some p := e do c`, at which point we have gone full circle to `let some p := e | c`. Hmm, maybe we shouldn't allow `doIfCond`s then.

**Comment by nomeata** [comment]
Can you easily and reliably recognize when `c` is obviously, syntactically not falling through?  Then a rule that would work rather well in practice could be for `(unless h : c do rhs); rest` that 

* `h : Not c`  is always available in `rhs`
* `h : c` is available in `rest` only if `rhs` is obviously not falling through.

This would capture the common idioms with `return`, `throw` and `break`. Downside: a bit confusing if the user thinks it's not falling through but lean doesn't think so.

**Comment by eric-wieser** [comment]
[#lean4 > while condition in function body @ 💬](https://leanprover.zulipchat.com/#narrow/channel/270676-lean4/topic/while.20condition.20in.20function.20body/near/608164717) has a similar request for `while`.

**Comment by sgraf812** [comment]
> Can you easily and reliably recognize when c is obviously, syntactically not falling through?

Yes, there is a [special `noFallthrough` flag in `ControlInfo`](https://github.com/leanprover/lean4/blob/a54f7e9c7ec212c629363121c0f977d07a8c4eff/src/Lean/Elab/Do/InferControlInfo.lean#L55). Still,

> `h : c` is available in rest only if rhs is obviously not falling through.

Is this really a good language feature? Should the presence of a terminal `return`/`break`/`continue` really govern whether `h : c` is in scope in `rest`? I'm torn. Similarly for `while h : ... do ...; rest`.

**Comment by nomeata** [comment]
I'm torn as well. The optimist in me says that it'll be exactly what almost everyone expects intuitively anyways, and it'll be natural even without explanation. The pessimist says the optimist is optimistic again.

Thinking too big for this PR: what if we had names that explicitly don't exist. Imagine in the then branch the elaborater would unconditionally shadow any outer `h` and resolve `g` to a helpful error message “it looks like you want to refer to the binder in line 42. This is not available here because the code guarded by unless could continue here. Ensure it ends with `return` or similar constructs”. Then I wouldn't be worried.
